### PR TITLE
Path resolution docs (current state only)

### PR DESCRIPTION
**Pull request by cameel**
Replaces #11105.
Related to #11138, #11036 and #9353.

This is a reworked version of #11105 that only describes the current state, without any proposed modifications. The goal is to get it merged and included in a release before we start changing how this stuff works.

To implement https://github.com/ethereum/solidity/pull/11105#pullrequestreview-617770242, I have created a separate section called "Path Resolution" in the side menu. The "Layout of a Solidity Source File" now only contains a short summary of the part needed to understand import paths.

The biggest change compared to the original docs is that any mention of absolute imports has been removed and now I only distinguish between "direct" and "relative" imports. I have also moved a lot of the extra details to notes/warnings to separate them from the main flow of the text and simplified them slightly where I could.

### CODE REVIEW

File: docs/layout-of-source-files.rst [hunk 1]
```diff
@@ -216,101 +216,62 @@ the code below creates new global symbols ``alias`` and ``symbol2`` which refere
...
-`Remix <https://remix.ethereum.org/>`_ provides an automatic remapping for
-GitHub and automatically retrieves the file over the network. You can import
-the iterable mapping as above,  e.g.
-
-::
+     import "./contracts/lib/token.sol" as token1;
+     import "../contracts/lib/token.sol" as token2;
 
-  import "github.com/ethereum/dapp-bin/library/iterable_mapping.sol" as it_mapping;
+  Unlike in case of a direct import, the compiler does assume that the import path is actually a path.
```
Comment (chriseth) [inline docs/layout-of-source-files.rst]: Isn't this complicating the isuse a little? Whta does it mean to assume it is a path? Isn't the rule just "if it starts with `./` or `../`, append it to the path and simplify"?

File: docs/layout-of-source-files.rst [hunk 2]
```diff
@@ -216,101 +216,62 @@ the code below creates new global symbols ``alias`` and ``symbol2`` which refere
...
+virtual filesystem, the compiler invokes the file loader.
+If you are using the command-line interface, the file loader simply assumes that the source unit
+name passed to it is a path and tries to load the file from the actual filesystem.
+The `JavaScript interface <https://github.com/ethereum/solc-js>`_ is a bit more flexible in that
+regard and allows the user to provide a callback to load the file.
+`The Remix IDE <https://remix.ethereum.org/>`_ uses this mechanism to allow files to be imported
+directly from github.
 
-.. note::
-    Always use relative imports like ``import "./filename.sol";`` and avoid
```
Comment (chriseth) [inline docs/layout-of-source-files.rst:236]: I would prefer this to still be in a prominent place.

File: docs/path-resolution.rst [hunk 3]
```diff
@@ -0,0 +1,690 @@
...
+This section aims to explain in detail how Solidity reconciles these requirements.
+
+.. index:: ! virtual filesystem, ! source unit name
+.. _virtual-filesystem:
+
+Virtual Filesystem
+==================
+
+The paths used in :ref:`imports <import>` in a general case do not have to be filesystem paths.
+The compiler maintains an internal database (virtual filesystem) where each compiled source unit is
```
Comment (chriseth) [inline docs/path-resolution.rst]: Why "compiled source unit" and not just "source unit"?

File: docs/path-resolution.rst [hunk 4]
```diff
@@ -0,0 +1,690 @@
...
+able to work with platform-specific paths to provide good user experience.
+This section aims to explain in detail how Solidity reconciles these requirements.
+
+.. index:: ! virtual filesystem, ! source unit name
+.. _virtual-filesystem:
+
+Virtual Filesystem
+==================
+
+The paths used in :ref:`imports <import>` in a general case do not have to be filesystem paths.
```
Comment (chriseth) [inline docs/path-resolution.rst]: The paths used in imports can also just be path fragments (if they are relative). Maybe we should start with the second sentence instead?

File: docs/path-resolution.rst [hunk 5]
```diff
@@ -0,0 +1,690 @@
...
+The compiler maintains an internal database (virtual filesystem) where each compiled source unit is
+assigned a unique *source unit name* which is an opaque and unstructured identifier.
+
+Source files can be placed in the virtual filesystem directly, using :ref:`Standard JSON interface
+<compiler-api>` and in this case source unit names can be anything.
+In most cases, however, the project files reside in the local filesystem and it is desirable for
+the compiler to be able to locate and load them automatically.
+To support this, the compiler passes all source unit names not found in the virtual filesystem to
```
Comment (chriseth) [inline docs/path-resolution.rst]: I think we should properly introduce the concept of "file loader" instead of just mentioning it in this sentence.

File: docs/path-resolution.rst [hunk 6]
```diff
@@ -0,0 +1,690 @@
...
+the file loader.
+In case of the command-line compiler the file loader attempts to interpret them as local paths.
```
Comment (chriseth) [inline docs/path-resolution.rst]: What is a local path? A path relative to the current working directory, or a path in the host filesystem?

File: docs/path-resolution.rst [hunk 7]
```diff
@@ -0,0 +1,690 @@
...
+The `JavaScript interface <https://github.com/ethereum/solc-js>`_ is a bit more flexible in that
+regard and allows the user to provide a callback, which can interpret the source unit name in an
+arbitrary way.
+For example load it from the network if it is an URL.
```
Comment (chriseth) [inline docs/path-resolution.rst]: Maybe: "Remix uses this to load files from github."

File: docs/path-resolution.rst [hunk 8]
```diff
@@ -0,0 +1,690 @@
...
+               }
+           },
+           "settings": {"outputSelection": {"*": { "*": ["metadata", "evm.bytecode"]}}}
+       }
+
+   The value specified in ``urls`` does not affect the source unit name and is not included in
+   contract metadata.
+   It is only passed to the file loader and used to locate the file.
+
+   As the name of the attribute implies, the value could be an URL if supported by the loader.
```
Comment (chriseth) [inline docs/path-resolution.rst]:
```suggestion
   As the name of the attribute implies, the value could be a URL if supported by the loader.
```

Comment (chriseth) [reply docs/path-resolution.rst]: Maybe clarify: "http URLs"?

File: docs/path-resolution.rst [hunk 9]
```diff
@@ -0,0 +1,690 @@
...
+
+   .. index:: ! import; path
+
+#. **import statement**
+
+   The ``import`` statement requests a module from the compiler and allows to access certain symbols
+   from that module.
+   We will refer to the path used in the statement as *import path*.
+
+   The import path is translated into a source unit name and then the compiler uses the name
```
Comment (chriseth) [inline docs/path-resolution.rst]:
```suggestion
   The import path is translated into a source unit name by combining it with the name of the importing source unit and then the compiler uses the name
```

File: docs/path-resolution.rst [hunk 10]
```diff
@@ -0,0 +1,690 @@
...
+   to look up the file in its virtual filesystem.
+   The are two types of imports, each with different rules for this translation:
+   :ref:`direct imports <direct-imports>` let you specify the full source unit name while in
```
Comment (chriseth) [inline docs/path-resolution.rst]: Why not "absolute"?

File: docs/path-resolution.rst [hunk 11]
```diff
@@ -0,0 +1,690 @@
...
+   .. index:: ! import; path
+
+#. **import statement**
+
+   The ``import`` statement requests a module from the compiler and allows to access certain symbols
+   from that module.
+   We will refer to the path used in the statement as *import path*.
+
+   The import path is translated into a source unit name and then the compiler uses the name
+   to look up the file in its virtual filesystem.
```
Comment (chriseth) [inline docs/path-resolution.rst]: If it is not found, the file loader is invoked to retrieve it.

File: docs/path-resolution.rst [hunk 12]
```diff
@@ -0,0 +1,690 @@
...
+   The are two types of imports, each with different rules for this translation:
+   :ref:`direct imports <direct-imports>` let you specify the full source unit name while in
+   :ref:`relative imports <relative-imports>` part of it comes from the source unit name of the
+   importing file.
+
+   .. index:: standard input, stdin, <stdin>
+
+#. **Standard input**
+
+   The last way to provide the source is by sending it to compiler's :ref:`standard input
```
Comment (chriseth) [inline docs/path-resolution.rst]: Only for the commandline compiler.

File: docs/path-resolution.rst [hunk 13]
```diff
@@ -0,0 +1,690 @@
...
+The import path translates directly to a source unit name without normalization of any kind:
+
+::
+
+    import "/project/lib/../lib///math.sol"; // source unit name: /project/lib/../lib///math.sol
+    import "lib/../lib///math.sol";          // source unit name: lib/../lib///math.sol
+
+In the above you might expect the source unit names to be ``/project/lib/math.sol`` and
+``lib/math.sol`` respectively but this is not the case.
+For direct imports the source unit name is exactly what is stated in the import.
```
Comment (chriseth) [inline docs/path-resolution.rst]: Unless remappings are used.

File: docs/path-resolution.rst [hunk 14]
```diff
@@ -0,0 +1,690 @@
...
+When the source is provided via Standard JSON interface each of these names can actually be
+associated with different content.
+
+When the source is not available in the virtual filesystem, the compiler passes the source unit name
+to the file loader.
+The default loader will attempt to use it as a path and look up the file on disk.
```
Comment (chriseth) [inline docs/path-resolution.rst]: is this true for all binaries?

File: docs/path-resolution.rst [hunk 15]
```diff
@@ -0,0 +1,690 @@
...
+    .. code-block:: solidity
+       :caption: /project/lib/math.sol
+
+       import "/project/lib/util.sol" as util1; // source unit name: /project/lib/util.sol
+       import "./util.sol" as util2;            // source unit name: /project/lib/util.sol
+       import "util.sol" as util3;              // source unit name: util.sol
+
+    In the situation above the first and the second import are equivalent and refer to the same
+    source unit in the virtual filesystem.
+    The compiler will recognize that the source has already been compiled when it encounters
```
Comment (chriseth) [inline docs/path-resolution.rst]:
```suggestion
    The compiler will recognize that the source has already been loaded when it encounters
```

File: docs/path-resolution.rst [hunk 16]
```diff
@@ -0,0 +1,690 @@
...
+    ``./util.sol`` and will not try to compile it again.
```
Comment (chriseth) [inline docs/path-resolution.rst]:
```suggestion
    ``./util.sol`` and will not try to load it again.
```

File: docs/path-resolution.rst [hunk 17]
```diff
@@ -0,0 +1,690 @@
...
+.. code-block:: solidity
+    :caption: lib/math.sol
+
+    import "./util.sol" as util;    // source unit name: lib/util.sol
+    import "../token.sol" as token; // source unit name: token.sol
+
+.. note::
+
+    Do not confuse relative imports with relative paths.
+    Both ``util.sol`` and ``./util.sol`` specify relative paths on disk but these paths are treated
```
Comment (chriseth) [inline docs/path-resolution.rst:297]: But this depends on the compiler binary. This section is rather confusing - maybe  we should clarify it as early as possible and talk about "filesystem paths relative to the current working directory"

File: docs/path-resolution.rst [hunk 18]
```diff
@@ -0,0 +1,690 @@
...
+    get resolved into ``/project/lib/util.sol`` by the file loader.
+    When the loader returns the source, the compiler will compile it but still place it under
+    ``util.sol`` and not ``/project/lib/util.sol`` in the virtual filesystem.
+
+Unlike in direct imports, the paths used in relative imports do get normalized.
+The normalization rules are the same as for UNIX paths, namely:
+
+- All the ``./`` segments are removed.
+- Every ``../``  segment backtracks one level up in the hierarchy.
+- Multiple slashes are squashed into a single one.
```
Comment (chriseth) [inline docs/path-resolution.rst:328]: Is this done to the full path? Maybe we should change that and only modify the initial segment.

File: docs/path-resolution.rst [hunk 19]
```diff
@@ -0,0 +1,690 @@
...
+    import "./util//util.sol";          // source unit name: lib/util/util.sol
+    import "../util/../array/util.sol"; // source unit name: array/util.sol
+
+.. warning::
+
+    The root of the virtual filesystem is an empty path, not ``/``.
+    This matters when the ``../`` segments go beyond the root.
+    In UNIX paths such segments are ignored and for example ``/../../`` is
+    equivalent to just ``/``.
+    In the virtual filesystem the rule similar but the result is an empty path instead.
```
Comment (chriseth) [inline docs/path-resolution.rst]:
```suggestion
    In the virtual filesystem the rule is similar but the result is an empty path instead.
```

File: docs/path-resolution.rst [hunk 20]
```diff
@@ -0,0 +1,690 @@
...
+.. code-block:: solidity
+    :caption: lib/contract.sol
+
+    import "./util/./util.sol";         // source unit name: lib/util/util.sol
+    import "./util//util.sol";          // source unit name: lib/util/util.sol
+    import "../util/../array/util.sol"; // source unit name: array/util.sol
+
+.. warning::
+
+    The root of the virtual filesystem is an empty path, not ``/``.
```
Comment (chriseth) [inline docs/path-resolution.rst:341]: Maybe rephrase: An initial `/` is not required (and also not recommended) for specifying absolute paths in the virtual filesystem.

File: docs/path-resolution.rst [hunk 21]
```diff
@@ -0,0 +1,690 @@
...
+        :caption: /project/lib/contract.sol
+
+        import "../util.sol";          // source unit name: /project/util.sol
+        import "../../util.sol";       // source unit name: /util.sol
+        import "../../../util.sol";    // source unit name: util.sol
+        import "../../../../util.sol"; // source unit name: util.sol
+
+.. note::
+
+    The parent source unit name is **not** normalized.
```
Comment (chriseth) [inline docs/path-resolution.rst]: "parent source unit name" is not properly defined. Maybe just use "importing source unit name"?

File: docs/path-resolution.rst [hunk 22]
```diff
@@ -0,0 +1,690 @@
...
+    The same effect can be achieved in a more reliable way by using direct imports with
+    :ref:`base path <base-path>` and :ref:`import remapping <import-remapping>`.
+
+.. index:: ! base path
+.. _base-path:
+
+Base Path
+=========
+
+Base path is the directory that contains all the files belonging to the project being compiled.
```
Comment (chriseth) [inline docs/path-resolution.rst]: I'm not sure this is a good definition.

File: docs/path-resolution.rst [hunk 23]
```diff
@@ -0,0 +1,690 @@
...
+The default assumption when using the command-line interface is that the compiler is being invoked
+from the base path.
+This way direct imports can use paths relative to project root and be successfully resolved by
```
Comment (chriseth) [inline docs/path-resolution.rst]:
```suggestion
This way direct imports can use paths relative to the project root and be successfully resolved by
```

File: docs/path-resolution.rst [hunk 24]
```diff
@@ -0,0 +1,690 @@
...
+By setting base path to ``/project`` you could then have the compiler find them in
+``/project/dapp-bin/library``
+
+The remappings can depend on a context, which allows you to configure packages to import,
+e.g. different versions of a library of the same name.
+
+.. warning::
+
+    Information about used remappings is stored in contract metadata so, while they let you avoid
+    changing the source, they cannot be used to achieve reproducible builds.
```
Comment (chriseth) [inline docs/path-resolution.rst]: This sounds like builds that use remappings are not reproducible.

File: docs/path-resolution.rst [hunk 25]
```diff
@@ -0,0 +1,690 @@
...
+    The metadata hash embedded in the bytecode will not be the same if you perform import remapping.
+
+Path remappings have the form of ``context:prefix=target``.
```
Comment (chriseth) [inline docs/path-resolution.rst]: Is a path remapping something else than an import remapping?

File: docs/path-resolution.rst (hunk 10 above)
Comment (cameel) [reply docs/path-resolution.rst]: Because when I see absolute imports, I think about absolute paths in the real filesystem and I really want to avoid that connotation. Thinking about absolute imports was really giving me the wrong idea about how this works when I was trying to understand this initially. I think we should have a different word to refer to them.

File: docs/path-resolution.rst (hunk 6 above)
Comment (cameel) [reply docs/path-resolution.rst]: The latter.

I think the text feels very repetitive and for some concepts I, unfortunately, have very few synonyms I can use without making things ambiguous. But I can go back to filesystem paths here.

File: docs/path-resolution.rst (hunk 12 above)
Comment (cameel) [reply docs/path-resolution.rst]: Yeah, but the whole section is written from the CLI standpoint and I'm only mentioning the JS interface as an aside in a few places. Should I change that?

File: docs/path-resolution.rst (hunk 13 above)
Comment (cameel) [reply docs/path-resolution.rst]: I wanted to keep it simple here and only expand on it later.

File: docs/path-resolution.rst (hunk 18 above)
Comment (cameel) [reply docs/path-resolution.rst:328]: Only the part that's in the import statement. The parent path it's combined with is not normalized. This is explained in a note later in the text.

Generally in this rewrite I tried to keep it simple in the initial explanations and only went into detail later in the text or in note/warning boxes.

File: docs/path-resolution.rst (hunk 20 above)
Comment (cameel) [reply docs/path-resolution.rst:341]: But this would sound like it's optional and the two are equivalent. The empty path does work like the root for us. See for example what happens when you go beyond the root - you get to `""` not `/`.

File: docs/path-resolution.rst (hunk 25 above)
Comment (cameel) [reply docs/path-resolution.rst]: It's the same. I'm calling it "import remapping" because it really only applies to paths used in imports. Initially I was confused by "path remapping" and thought that it applies to all paths. I.e. those used on the CLI, in JSON, etc. I think the new name clarifies this a bit.

File: docs/layout-of-source-files.rst (hunk 1 above)
Comment (cameel) [reply docs/layout-of-source-files.rst]: It assumes it's structured like a path, e.g. expects path separators or squashes `../` segments. It's even processed internally using `boost::filesystem`.

File: docs/layout-of-source-files.rst (hunk 2 above)
Comment (cameel) [reply docs/layout-of-source-files.rst:236]: Done.

File: docs/path-resolution.rst (hunk 3 above)
Comment (cameel) [reply docs/path-resolution.rst]: I did it mostly to avoid repetition. Changed.

File: docs/path-resolution.rst (hunk 4 above)
Comment (cameel) [reply docs/path-resolution.rst]: Done.

File: docs/path-resolution.rst (hunk 5 above)
Comment (cameel) [reply docs/path-resolution.rst]: This was a great idea. I added a section and introduced separate terms for the callback (*import callback*) and the built-in loader (*Host Filesystem Loader*). I ended up rewriting the rest of text a bit to match but I think that made it more consistent and now I don't have to mention the JS interface at every step.

File: docs/path-resolution.rst (hunk 7 above)
Comment (cameel) [reply docs/path-resolution.rst]: I originally did not want to repeat it (it's already mentioned in the section about URLs). But after adding the section on the callback I moved that mention here.

File: docs/path-resolution.rst (hunk 8 above)
Comment (cameel) [reply docs/path-resolution.rst]: Fixed.

> Maybe clarify: "http URLs"?

It's not just HTTP thought. Any protocol that the callback can handle is allowed. For example Remix can now handle IPFS.

File: docs/path-resolution.rst (hunk 9 above)
Comment (cameel) [reply docs/path-resolution.rst]: Done.

File: docs/path-resolution.rst (hunk 12 above)
Comment (cameel) [reply docs/path-resolution.rst]: Changed.

File: docs/path-resolution.rst (hunk 13 above)
Comment (cameel) [reply docs/path-resolution.rst]: Changed.

File: docs/path-resolution.rst (hunk 14 above)
Comment (cameel) [reply docs/path-resolution.rst]: Well, technically yes because the default loader is present in all binaries but some binaries just do not use it :)

Anyway, this is now reworded after I added a section on the callback.

File: docs/path-resolution.rst (hunk 15 above)
Comment (cameel) [reply docs/path-resolution.rst]: Fixed.

File: docs/path-resolution.rst (hunk 16 above)
Comment (cameel) [reply docs/path-resolution.rst]: Fixed.

File: docs/path-resolution.rst (hunk 19 above)
Comment (cameel) [reply docs/path-resolution.rst]: Fixed.

File: docs/path-resolution.rst (hunk 21 above)
Comment (cameel) [reply docs/path-resolution.rst]: Done.

File: docs/path-resolution.rst (hunk 22 above)
Comment (cameel) [reply docs/path-resolution.rst]: Reworded.

I think this is basically what it boils down to in the current implementation - if you use non-empty base path everything has to sit in a single directory.

File: docs/path-resolution.rst (hunk 23 above)
Comment (cameel) [reply docs/path-resolution.rst]: Fixed.

File: docs/path-resolution.rst (hunk 24 above)
Comment (cameel) [reply docs/path-resolution.rst]: Reworded.

File: docs/path-resolution.rst (hunk 25 above)
Comment (cameel) [reply docs/path-resolution.rst]: I now see that you just pointed out a single place where I still used the old term. Fixed.

**Comment by cameel** [comment]
I have just pushed a new version with all the points above addressed. All the changes are in fixups.

I'm going to close it now and split in two parts. One will go into a separate PR with the excessive examples and explanations removed. The other will become a blog post.

File: docs/path-resolution.rst (hunk 17 above)
Comment (cameel) [reply docs/path-resolution.rst:297]: I'm leaving it here for now but I'll probably rewrite it for the blog post anyway. In #11442 I removed it almost completely and just added a short note.

File: docs/path-resolution.rst (hunk 20 above)
Comment (cameel) [reply docs/path-resolution.rst:341]: I have rewritten this section completely in #11442.

**Comment by cameel** [comment]
Closing in favor of #11442.
